Copyedits (#2)
R-ArcGIS/r-bridge-site: geoprocessing-tools.qmd

![](images/gp-tool-properties.png){width=50%}

If the direction is `Input`, it will be contained in `in_params`. Likewise, 
if the direction is `Output`, it will be contained in `out_params`. 

## Using parameters

> Values can be extracted from the `in_params` and `out_params` lists by the name or position of the parameter. It is highly recommended to use name-based indexing rather than position-based indexing for clarity.

Reviewer: ```suggestion
Values can be extracted from the `in_params` and `out_params` lists by the name or position of the parameter. It is strongly recommended to use name-based indexing rather than position-based indexing for clarity.
```

:::{.callout-caution}
The name of the parameter must match the value in the **Name** (_not_ Label) column in the `Parameters` tab of the Tool Properties.
:::

> This is an example of a function to parse a parameter named `date`:

Reviewer: ```suggestion
This is an example of a function to parse a `date` parameter with the name `date`:
```

```

:::{.callout-note}
To filter or select columns from a dataset, consider using the `fields` and `where_clause` arguments of `arc.select()` to reduce the amount of data read into memory. 

`arc.select()` returns a `data.frame` with fields and a special geometry column which is incompatible with sf. Use `arc.data2sf()` to convert it to an sf object.
:::

### Writing a Feature Class

It is quite common to write the results of an analysis to a file geodatabase. This can be done with `arc.write()`. A Feature Class type parameter can be used in the `out_params` list object to capture a user-defined output path.

`arc.write()` requires two arguments. The first is the output path to write to and the second is the object to write. The accepted types of objects are `data.frame`, `sf`, `RasterLayer` or `RasterBrick`. 

:::{.callout-note}

> As of Nov 2023, support for `{terra}` is in development.

Reviewer: Support for `{terra}` is planned to accompany the ArcGIS Pro version 3.3 release.

### Installing Required Packages 

> When sharing R-based GP tools with other users, they may not have the packages that the script tool needs to execute code. In this case, required packages should be automatically installed the first time the script tool is executed.

Reviewer: I'm wondering if we should change the word "should" to "can be." "Should" feels a bit forceful and  there are other ways of accomplishing it as well.

> The script tool can check to see if the required package is installed and if not, install it. Below is a helper function to include at the top of the script if there are multiple packages to check:

Reviewer: >  package is installed and if not, install it.

should become package is installed and, if not, install it.

}
```

> This function uses `requireNamespace()` which attempts to load the provided package. If it succeeds, it returns `TRUE`, and if not it returns `FALSE`. The function checks to see if `FALSE` is returned and if so, installs the package and prints an informative message. 

Reviewer: ```suggestion
This function uses `requireNamespace()` which attempts to load the provided package. If it succeeds, it returns `TRUE`, and if not it returns `FALSE`. The function checks to see if `FALSE` is returned and, if so, installs the package and prints an informative message. 
```
